File: src/main/java/com/financemanager/controllers/CategoriesController.java
package com.financemanager.controllers;

import com.financemanager.service.CategoryManager;

import javafx.fxml.FXML;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.event.ActionEvent;
import javafx.stage.FileChooser;
import javafx.stage.Stage;

import java.io.File;
import java.util.ArrayList;

public class CategoriesController {

    @FXML private Button importCategoriesButton = new Button();
    @FXML private Button closeButton = new Button();

    @FXML ListView<String> categoryList;
    @FXML ListView<String> keywordList;

    private final CategoryManager categoryManagerInstance = CategoryManager.getInstance();

    @FXML
    private void handleCloseCategories(ActionEvent event) {
        Stage stage = (Stage) ((Button) event.getSource()).getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleImportCategories(ActionEvent event) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Import Categories File");
        FileChooser.ExtensionFilter extFilter = new FileChooser.ExtensionFilter("Data Files (*.dat)", "*.dat");
        fileChooser.getExtensionFilters().addAll(extFilter);

        Stage stage = (Stage) ((Node) event.getSource()).getScene().getWindow();
        File file = fileChooser.showOpenDialog(stage);

        if (file != null) {
            categoryManagerInstance.importFromFile(file);
            refreshCategoryList();
        }
    }

    private void refreshCategoryList() {
        categoryList.getItems().clear();
        keywordList.getItems().clear();
        categoryList.getItems().addAll(categoryManagerInstance.getOrderedCategories());
    }

    @FXML
    public void initialize() {
        categoryManagerInstance.syncOrder();

        categoryList.getItems().addAll(categoryManagerInstance.getOrderedCategories());

        categoryList.getSelectionModel().selectedItemProperty().addListener((observable, oldValue, selectedCat)-> {
            keywordList.getItems().clear();
            if (selectedCat != null) {
                ArrayList<String> keys = categoryManagerInstance.getCategoriesMap().get(selectedCat);
                if (keys != null) {
                    keywordList.getItems().addAll(keys);
                }
            }
        });
        setupCategoryContextMenu();
        setupKeywordContextMenu();
    }

    private void handleAddCategory() {
        TextInputDialog dialog = new TextInputDialog();
        dialog.setTitle("New Group");
        dialog.setHeaderText("Create a new category group");
        dialog.showAndWait().ifPresent (name -> {
            String finalName = name.trim();
            if (!finalName.isEmpty()) {
                if (!categoryManagerInstance.getCategoriesMap().containsKey(finalName)) {
                    categoryManagerInstance.addCategory(finalName);
                    categoryList.getItems().add(finalName);
                    categoryManagerInstance.save();
                } else {
                    Alert alert = new Alert(Alert.AlertType.ERROR, "Group Already Exists");
                    alert.showAndWait();
                }
            }
        });
    }

    private void handleEditCategory(String oldName) {
        TextInputDialog dialog = new TextInputDialog(oldName);
        dialog.setTitle("Edit Category");
        dialog.setHeaderText("Rename " + oldName + " to:");
        dialog.showAndWait().ifPresent (newName -> {
            String finalName = newName.trim();
            if (!finalName.isEmpty() && !finalName.equals(oldName)) {
                categoryManagerInstance.renameCategory(oldName, finalName);
                int index = categoryList.getItems().indexOf(oldName);
                categoryList.getItems().set(index, newName);
                categoryManagerInstance.save();
            }
        });
    }

    private void handleRemoveCategory(String name) {
        Alert alert = new Alert(Alert.AlertType.CONFIRMATION, "Delete group " + name + "?", ButtonType.YES, ButtonType.NO);
        if (alert.showAndWait().get() == ButtonType.YES) {
            categoryManagerInstance.removeCategory(name);
            categoryList.getItems().remove(name);
            keywordList.getItems().clear();
            categoryManagerInstance.save();
        }
    }

    // --- MÉTODOS DE KEYWORD (DIREITA) --- [cite: 2026-01-24]

    private void handleAddKeyword() {
        String selectedCat = categoryList.getSelectionModel().getSelectedItem();
        if (selectedCat == null) {
            new Alert(Alert.AlertType.WARNING, "Select a Group on the left first!").showAndWait();
            return;
        }
        TextInputDialog dialog = new TextInputDialog();
        dialog.setTitle("Add Keyword");
        dialog.setHeaderText("Add keyword for " + selectedCat);
        dialog.showAndWait().ifPresent (key -> {
            String finalKey = key.trim();
            if (!finalKey.isEmpty()) {
                categoryManagerInstance.addKeyword(selectedCat, finalKey);
                keywordList.getItems().add(finalKey);
                categoryManagerInstance.save();
            }
        });
    }

    private void handleEditKeyword(String oldKey) {
        String selectedCat = categoryList.getSelectionModel().getSelectedItem();
        if (selectedCat == null) {
            return;
        }
        TextInputDialog dialog = new TextInputDialog(oldKey);
        dialog.setTitle("Edit Keyword");
        dialog.setHeaderText("Rename keyword " + oldKey + ":");
        dialog.showAndWait().ifPresent(newKey -> {
            String finalKey = newKey.trim();
            if (!finalKey.isEmpty() && !finalKey.equals(oldKey)) {
                ArrayList<String> keys = categoryManagerInstance.getCategoriesMap().get(selectedCat);
                int index = keys.indexOf(oldKey);
                if (index != -1) {
                    keys.set(index, finalKey);
                    int visualIndex = keywordList.getItems().indexOf(oldKey);
                    keywordList.getItems().set(visualIndex, finalKey);
                    categoryManagerInstance.save();
                }
            }
        });
    }

    private void handleRemoveKeyword(String keyword) {
        String selectedCat = categoryList.getSelectionModel().getSelectedItem();
        if (selectedCat == null) {
            return;
        }
        categoryManagerInstance.getCategoriesMap().get(selectedCat).remove(keyword);
        keywordList.getItems().remove(keyword);
        categoryManagerInstance.save();
    }

    private void setupCategoryContextMenu() {
        ContextMenu contextMenu = new ContextMenu();
        MenuItem add = new MenuItem("Add Category");
        MenuItem edit = new MenuItem("Edit");
        MenuItem remove = new MenuItem("Remove");

        add.setOnAction(e -> handleAddCategory());

        categoryList.setContextMenu(contextMenu);

        categoryList.setOnContextMenuRequested(e -> {
            String selected = categoryList.getSelectionModel().getSelectedItem();
            contextMenu.getItems().clear();
            if (selected != null) {
                edit.setOnAction(ev -> handleEditCategory(selected));
                remove.setOnAction(ev -> handleRemoveCategory(selected));
                contextMenu.getItems().addAll(add, edit, remove);
            } else {
                contextMenu.getItems().add(add);
            }
        });
    }

    private void setupKeywordContextMenu() {
        ContextMenu contextMenu = new ContextMenu();
        MenuItem add = new MenuItem("Add Keyword");
        MenuItem edit = new MenuItem("Edit");
        MenuItem remove = new MenuItem("Remove");

        add.setOnAction(e -> handleAddKeyword());

        keywordList.setContextMenu(contextMenu);

        keywordList.setOnContextMenuRequested(e -> {
            String selected = keywordList.getSelectionModel().getSelectedItem();
            contextMenu.getItems().clear();
            if (selected != null) {
                edit.setOnAction(ev -> handleEditKeyword(selected));
                remove.setOnAction(ev -> handleRemoveKeyword(selected));
                contextMenu.getItems().addAll(add, edit, remove);
            } else {
                contextMenu.getItems().add(add);
            }
        });
    }

}

File: src/main/kotlin/com/financemanager/controllers/CategoriesController.kt
package com.financemanager.controllers

import com.financemanager.service.CategoryManager
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.Alert
import javafx.scene.control.Button
import javafx.scene.control.ButtonType
import javafx.scene.control.ContextMenu
import javafx.scene.control.ListView
import javafx.scene.control.MenuItem
import javafx.scene.control.TextInputDialog

class CategoriesController {

    @FXML var importCategoriesButton: Button = Button()
    @FXML var closeButton: Button = Button()

    @FXML lateinit var categoryList: ListView<String>
    @FXML lateinit var keywordList: ListView<String>

    @FXML
    fun handleCloseCategories(event: ActionEvent) {
        val stage = (event.source as Button).scene.window as javafx.stage.Stage
        stage.close()
    }

    @FXML
    fun handleImportCategories(event: ActionEvent) {
        val fileChooser = javafx.stage.FileChooser()
        fileChooser.title = "Import Categories File"
        val extFilter = javafx.stage.FileChooser.ExtensionFilter("Data Files (*.dat)", "*.dat")
        fileChooser.extensionFilters.add(extFilter)

        val stage = (event.source as javafx.scene.Node).scene.window as javafx.stage.Stage
        val file = fileChooser.showOpenDialog(stage)

        if (file != null) {
            CategoryManager.getInstance().importCategories(file)
            refreshCategoryList()
            println("Ficheiro importado com sucesso: ${file.name}")
        }
    }

    private fun refreshCategoryList() {
        categoryList.items.clear()
        keywordList.items.clear()
        categoryList.items.addAll(CategoryManager.getInstance().orderedCategories)
    }

    @FXML
    fun initialize() {
        println("DEBUG: CategoriesController initialized!")
        val catManager = CategoryManager.getInstance()
        catManager.syncOrder()

        categoryList.items.addAll(catManager.orderedCategories)

        // Sincronização das listas [cite: 2026-01-24]
        categoryList.selectionModel.selectedItemProperty().addListener { _, _, selectedCat ->
            keywordList.items.clear()
            if (selectedCat != null) {
                val keys = catManager.categoriesMap[selectedCat]
                if (keys != null) {
                    keywordList.items.addAll(keys)
                }
            }
        }

        // Configuração dos menus de contexto (Botão Direito) [cite: 2026-01-24]
        setupCategoryContextMenu()
        setupKeywordContextMenu()
    }

    // --- MÉTODOS DE CATEGORIA (ESQUERDA) --- [cite: 2026-01-24]

    private fun handleAddCategory() {
        val dialog = TextInputDialog()
        dialog.title = "New Group"
        dialog.headerText = "Create a new category group"
        dialog.showAndWait().ifPresent { name ->
            val finalName = name.trim()
            if (finalName.isNotEmpty()) {
                if (!CategoryManager.getInstance().categoriesMap.containsKey(finalName)) {
                    CategoryManager.getInstance().addCategory(finalName)
                    categoryList.items.add(finalName)
                    CategoryManager.getInstance().save()
                } else {
                    Alert(Alert.AlertType.ERROR, "Group already exists!").showAndWait()
                }
            }
        }
    }

    private fun handleEditCategory(oldName: String) {
        val dialog = TextInputDialog(oldName)
        dialog.title = "Edit Category"
        dialog.headerText = "Rename '$oldName' to:"
        dialog.showAndWait().ifPresent { newName ->
            val finalName = newName.trim()
            if (finalName.isNotEmpty() && finalName != oldName) {
                CategoryManager.getInstance().renameCategory(oldName, finalName)
                val index = categoryList.items.indexOf(oldName)
                categoryList.items[index] = finalName
                CategoryManager.getInstance().save()
            }
        }
    }

    private fun handleRemoveCategory(name: String) {
        val alert = Alert(Alert.AlertType.CONFIRMATION, "Delete group '$name'?", ButtonType.YES, ButtonType.NO)
        if (alert.showAndWait().get() == ButtonType.YES) {
            CategoryManager.getInstance().removeCategory(name)
            categoryList.items.remove(name)
            keywordList.items.clear()
            CategoryManager.getInstance().save()
        }
    }

    // --- MÉTODOS DE KEYWORD (DIREITA) --- [cite: 2026-01-24]

    private fun handleAddKeyword() {
        val selectedCat = categoryList.selectionModel.selectedItem
        if (selectedCat == null) {
            Alert(Alert.AlertType.WARNING, "Select a Group on the left first!").showAndWait()
            return
        }
        val dialog = TextInputDialog()
        dialog.title = "Add Keyword"
        dialog.headerText = "Add keyword for $selectedCat"
        dialog.showAndWait().ifPresent { key ->
            val finalKey = key.trim()
            if (finalKey.isNotEmpty()) {
                CategoryManager.getInstance().addKeyword(selectedCat, finalKey)
                keywordList.items.add(finalKey)
                CategoryManager.getInstance().save()
            }
        }
    }

    private fun handleEditKeyword(oldKey: String) {
        val selectedCat = categoryList.selectionModel.selectedItem ?: return
        val dialog = TextInputDialog(oldKey)
        dialog.title = "Edit Keyword"
        dialog.headerText = "Rename keyword '$oldKey':"
        dialog.showAndWait().ifPresent { newKey ->
            val finalKey = newKey.trim()
            if (finalKey.isNotEmpty() && finalKey != oldKey) {
                val keys = CategoryManager.getInstance().categoriesMap[selectedCat]
                val index = keys?.indexOf(oldKey) ?: -1
                if (index != -1) {
                    keys?.set(index, finalKey)
                    val visualIndex = keywordList.items.indexOf(oldKey)
                    keywordList.items[visualIndex] = finalKey
                    CategoryManager.getInstance().save()
                }
            }
        }
    }

    private fun handleRemoveKeyword(keyword: String) {
        val selectedCat = categoryList.selectionModel.selectedItem ?: return
        CategoryManager.getInstance().categoriesMap[selectedCat]?.remove(keyword)
        keywordList.items.remove(keyword)
        CategoryManager.getInstance().save()
    }

    // --- CONFIGURAÇÃO DOS MENUS DE CONTEXTO --- [cite: 2026-01-24]

    private fun setupCategoryContextMenu() {
        val contextMenu = ContextMenu()
        val add = MenuItem("Add Category").apply { setOnAction { handleAddCategory() } }
        val edit = MenuItem("Edit")
        val remove = MenuItem("Remove")

        categoryList.contextMenu = contextMenu

        // Antes de mostrar o menu, verificamos se há seleção [cite: 2026-01-24]
        categoryList.setOnContextMenuRequested {
            val selected = categoryList.selectionModel.selectedItem
            contextMenu.items.clear()
            if (selected != null) {
                edit.setOnAction { handleEditCategory(selected) }
                remove.setOnAction { handleRemoveCategory(selected) }
                contextMenu.items.addAll(add, edit, remove)
            } else {
                contextMenu.items.add(add)
            }
        }
    }

    private fun setupKeywordContextMenu() {
        val contextMenu = ContextMenu()
        val add = MenuItem("Add Keyword").apply { setOnAction { handleAddKeyword() } }
        val edit = MenuItem("Edit")
        val remove = MenuItem("Remove")

        keywordList.contextMenu = contextMenu

        keywordList.setOnContextMenuRequested {
            val selected = keywordList.selectionModel.selectedItem
            contextMenu.items.clear()
            if (selected != null) {
                edit.setOnAction { handleEditKeyword(selected) }
                remove.setOnAction { handleRemoveKeyword(selected) }
                contextMenu.items.addAll(add, edit, remove)
            } else {
                contextMenu.items.add(add)
            }
        }
    }
}